src/Projects/Modules/Crunchyroll/Clients/Module.Crunchyroll.Avalonia/Services/ImageDownlodService.cs

```diff
@@ -0,0 +1,68 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Threading.Tasks;
+using Avalonia.Media.Imaging;
+
+namespace Module.Crunchyroll.Avalonia.Services
+{
+    public class ImageDownlodService : IImageDownloadService, IDisposable
+    {
+        Dictionary<string, Stream> cache = new Dictionary<string, Stream>();
+
+        public void DownloadImageAsync(string url, Action<IBitmap> continueWith)
+        {
+            Task.Run(async () =>
+            {
+                if (!this.cache.TryGetValue(url, out var stream))
+                {
+                    stream = await this.DownloadImageAsync(url);
+                    this.cache.Add(url, stream);
+                }
+                var memoryStream = new MemoryStream();
+                stream.Seek(0, SeekOrigin.Begin);
+                await stream.CopyToAsync(memoryStream);
+                stream.Seek(0, SeekOrigin.Begin);
+                memoryStream.Seek(0, SeekOrigin.Begin);
+                
+                continueWith(new Bitmap(memoryStream));
+            });
+        }
+
+        private async Task<Stream> DownloadImageAsync(string url)
+        {
+            var request = WebRequest.Create(new Uri(url, UriKind.Absolute));
+            request.Timeout = -1;
+            using var response = await request.GetResponseAsync();
+            await using var responseStream = response.GetResponseStream();
+
+            if (responseStream is null)
+            {
+                throw new InvalidOperationException();
+            }
+
+            var memoryStream = new MemoryStream();
+
+            await responseStream.CopyToAsync(memoryStream);
+
+            memoryStream.Seek(0, SeekOrigin.Begin);
+
+            return memoryStream;
```

Comment:
Why the extra copying of responseStream?

Reply:
Because the `responseStream` is a `WebRequestResponseStream` and i don't want that the caller of the method needs to manage such a stream. A `MemoryStream` is more generic and versatile and don't require a further dependency and so the `WebRequestResponseStream` is only valid and managed in the context where it's created.